typescript/web/src/components/datasets/upsert-dataset-modal.tsx

```diff
@@ -102,8 +121,50 @@ export const UpsertDatasetModal = ({
         name: datasetName,
         workspaceSlug,
       },
-      refetchQueries: ["getDatasets"],
-      awaitRefetchQueries: true,
+      update: (cache, { data }) => {
+        if (data != null) {
+          const { createDataset } = data;
+          const datasetCacheResult = cache.readQuery<Pick<Query, "datasets">>({
+            query: getDatasetsQuery,
+            variables: { where: { workspaceSlug } },
+          });
+          if (datasetCacheResult?.datasets == null) {
+            throw new Error(`Error retrieving datasets`);
+          }
+
+          const { datasets } = datasetCacheResult;
+          const updatedDatasets = datasets.concat(createDataset);
+          cache.writeQuery({
+            query: getDatasetsQuery,
+            variables: { where: { workspaceSlug } },
+            data: { datasets: updatedDatasets },
+          });
+        } else {
+          throw new Error("Received null data during dataset creation");
+        }
+      },
+      refetchQueries: ["getPaginatedDatasets"],
+      optimisticResponse: {
+        createDataset: {
+          id: "",
```

Comment:
No ID? Won't this create bugs?

Reply:
Fixed

Reply:
No as the refetchQuery `getPaginatedDatasets` will override the missing ID